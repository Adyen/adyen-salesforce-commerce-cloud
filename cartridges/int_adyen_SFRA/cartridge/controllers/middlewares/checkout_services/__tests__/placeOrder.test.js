"use strict";

/* eslint-disable global-require */

var req;
var res;
var placeOrder = require('../placeOrder');
beforeEach(function () {
  var hooksHelper = require('*/cartridge/scripts/helpers/hooks');
  hooksHelper.mockImplementation(function () {
    return {};
  });
  jest.clearAllMocks();
  res = {
    json: jest.fn()
  };
  req = {
    session: {
      privacyCache: {
        get: jest.fn(),
        set: jest.fn()
      }
    },
    currentCustomer: {
      addressBook: true
    },
    locale: {
      id: 'nl_NL'
    }
  };
});
describe('Checkout Services', function () {
  it('should go to next middleware if payment instrument is not Adyen', function () {
    var _require = require('*/cartridge/controllers/middlewares/checkout_services/adyenCheckoutServices'),
      isNotAdyen = _require.isNotAdyen;
    isNotAdyen.mockImplementationOnce(jest.fn(function () {
      return true;
    }));
    var _require2 = require('*/cartridge/scripts/helpers/basketValidationHelpers'),
      validateProducts = _require2.validateProducts;
    var next = jest.fn();
    placeOrder(req, res, next);
    expect(next).toHaveBeenCalledTimes(1);
    expect(validateProducts).toHaveBeenCalledTimes(0);
  });
  it('should process payment if action is valid', function () {
    var _require3 = require('*/cartridge/controllers/middlewares/checkout_services/adyenCheckoutServices'),
      processPayment = _require3.processPayment;
    placeOrder.call({
      emit: jest.fn()
    }, req, res, jest.fn());
    expect(processPayment).toHaveBeenCalledTimes(1);
  });
  it('should not process payment and return json response at end of file when there is no action', function () {
    var adyenHelpers = require('*/cartridge/scripts/checkout/adyenHelpers');
    var COHelpers = require('*/cartridge/scripts/checkout/checkoutHelpers');
    adyenHelpers.handlePayments.mockImplementationOnce(function () {
      return {
        error: false
      };
    });
    placeOrder.call({
      emit: jest.fn()
    }, req, res, jest.fn());
    expect(COHelpers.sendConfirmationEmail).toBeCalledTimes(1);
    expect(res.json.mock.calls).toMatchSnapshot();
  });
  it('should attempt to cache orderNumber after order creation', function () {
    placeOrder.call({
      emit: jest.fn()
    }, req, res, jest.fn());
    expect(req.session.privacyCache.set.mock.calls[0][0]).toBe('currentOrderNumber');
    expect(req.session.privacyCache.set.mock.calls[0][1]).toBe('mocked_orderNo');
    expect(req.session.privacyCache.set.mock.calls[1][0]).toBe('currentOrderToken');
    expect(req.session.privacyCache.set.mock.calls[1][1]).toBe('mocked_orderToken');
  });
});