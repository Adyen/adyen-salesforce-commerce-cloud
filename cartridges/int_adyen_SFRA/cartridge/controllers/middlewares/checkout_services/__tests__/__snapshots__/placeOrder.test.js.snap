// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Place Order should complete if viewData contains csrfError 1`] = `
Array [
  Array [
    "route:Complete",
    Object {
      "currentCustomer": Object {
        "addressBook": true,
      },
      "locale": Object {
        "id": "NL",
      },
      "session": Object {
        "privacyCache": Object {
          "get": [MockFunction],
          "set": [MockFunction],
        },
      },
    },
    Object {
      "getViewData": [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": Object {
              "csrfError": true,
            },
          },
        ],
      },
      "json": [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      },
    },
  ],
]
`;

exports[`Place Order should fail if billing address is empty 1`] = `
Array [
  Array [
    Object {
      "error": true,
      "errorMessage": "mocked_error.no.billing.address",
      "errorStage": Object {
        "stage": "payment",
        "step": "billingAddress",
      },
    },
  ],
]
`;

exports[`Place Order should fail if calculated payment total has error 1`] = `
Array [
  Array [
    Object {
      "error": true,
      "errorMessage": "mocked_error.technical",
    },
  ],
]
`;

exports[`Place Order should fail if created order is empty 1`] = `
Array [
  Array [
    Object {
      "error": true,
      "errorMessage": "mocked_error.technical",
    },
  ],
]
`;

exports[`Place Order should fail if fraudDetection status returns "fail" 1`] = `
Array [
  Array [
    Object {
      "cartError": true,
      "error": true,
      "errorMessage": "mocked_error.technical",
      "redirectUrl": "[\\"Error-ErrorCode\\",\\"err\\",\\"mocked_fraud_fail\\"]",
    },
  ],
]
`;

exports[`Place Order should fail if order validation status is error 1`] = `
Array [
  Array [
    Object {
      "error": true,
      "errorMessage": "mocked_order_validation_message",
    },
  ],
]
`;

exports[`Place Order should fail if placeOrder has error 1`] = `
Array [
  Array [
    Object {
      "error": true,
      "errorMessage": "mocked_error.technical",
    },
  ],
]
`;

exports[`Place Order should fail if shipping address is empty 1`] = `
Array [
  Array [
    Object {
      "error": true,
      "errorMessage": "mocked_error.no.shipping.address",
      "errorStage": Object {
        "stage": "shipping",
        "step": "address",
      },
    },
  ],
]
`;

exports[`Place Order should fail if there is error on payment handle 1`] = `
Array [
  Array [
    Object {
      "error": true,
      "errorMessage": "mocked_error.payment.not.valid",
    },
  ],
]
`;

exports[`Place Order should handle threeDS2 1`] = `
Array [
  Array [
    Object {
      "continueUrl": "[\\"Adyen-Adyen3DS2\\",\\"resultCode\\",\\"mocked_threeDS2_resultCode\\",\\"orderNo\\",\\"mocked_orderNo\\"]",
      "error": false,
      "order": Object {
        "getPaymentInstruments": [MockFunction] {
          "calls": Array [
            Array [
              "mocked_method",
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": Object {
                "0": Object {
                  "custom": Object {
                    "adyenAction": "mocked_action",
                    "adyenMD": "mocked_adyen_MD",
                    "adyenPaymentData": "mocked_adyen_payment_data",
                    "adyenRedirectURL": "https://some_mocked_url/signature",
                  },
                },
                "iterator": [Function],
                "toArray": [MockFunction] {
                  "calls": Array [
                    Array [],
                  ],
                  "results": Array [
                    Object {
                      "type": "return",
                      "value": Array [
                        Object {
                          "custom": Object {
                            "adyenAction": "mocked_action",
                            "adyenMD": "mocked_adyen_MD",
                            "adyenPaymentData": "mocked_adyen_payment_data",
                            "adyenRedirectURL": "https://some_mocked_url/signature",
                          },
                        },
                      ],
                    },
                  ],
                },
              },
            },
          ],
        },
        "orderNo": "mocked_orderNo",
        "orderToken": "mocked_orderToken",
      },
    },
  ],
]
`;

exports[`Place Order should redirect with authorize3d 1`] = `
Array [
  Array [
    Object {
      "continueUrl": "[\\"Adyen-Adyen3D\\",\\"IssuerURL\\",\\"mocked_url\\",\\"PaRequest\\",\\"mocked_PaReq\\",\\"MD\\",\\"mocked_MD\\",\\"merchantReference\\",\\"mocked_orderNo\\",\\"signature\\",null]",
      "error": false,
    },
  ],
]
`;

exports[`Place Order should redirect without authorize3d 1`] = `
Array [
  Array [
    Object {
      "continueUrl": "[\\"Adyen-Redirect\\",\\"merchantReference\\",\\"mocked_orderNo\\",\\"signature\\",null]",
      "error": false,
    },
  ],
]
`;

exports[`Place Order should return error if privacy cache detect fraud 1`] = `
Array [
  Array [
    Object {
      "cartError": true,
      "error": true,
      "errorMessage": "mocked_error.technical",
      "redirectUrl": "[\\"Error-ErrorCode\\",\\"err\\",\\"01\\"]",
    },
  ],
]
`;

exports[`Place Order should return error when basket is empty 1`] = `
Array [
  Array [
    Object {
      "cartError": true,
      "error": true,
      "fieldErrors": Array [],
      "redirectUrl": "[\\"Cart-Show\\"]",
      "serverErrors": Array [],
    },
  ],
]
`;

exports[`Place Order should return json response 1`] = `
Array [
  Array [
    Object {
      "continueUrl": "[\\"Order-Confirm\\"]",
      "error": false,
      "orderID": "mocked_orderNo",
      "orderToken": "mocked_orderToken",
    },
  ],
]
`;
